From 486110e2564cd9b677c485ec87b5df1bcb5750f2 Mon Sep 17 00:00:00 2001
From: rpm-build <rpm-build>
Date: Wed, 29 Aug 2018 16:20:27 -0700
Subject: [PATCH 184/184] set iscsid.safe_logout to Yes by default

---
 etc/iscsid.conf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/etc/iscsid.conf b/etc/iscsid.conf
index cbf2c05aff59..1af8ed2732de 100644
--- a/etc/iscsid.conf
+++ b/etc/iscsid.conf
@@ -25,7 +25,7 @@ iscsid.startup = /bin/systemctl start iscsid.socket iscsiuio.socket
 
 # Check for active mounts on devices reachable through a session
 # and refuse to logout if there are any.  Defaults to "No".
-# iscsid.safe_logout = Yes
+iscsid.safe_logout = Yes
 
 #############################
 # NIC/HBA and driver settings
-- 
2.14.4

